File: pgpainless-sop/src/main/java/org/pgpainless/sop/DearmorImpl.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.sop;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import org.bouncycastle.openpgp.PGPUtil;
import org.bouncycastle.util.io.Streams;
import sop.Ready;
import sop.exception.SOPGPException;
import sop.operation.Dearmor;

import javax.annotation.Nonnull;

/**
 * Implementation of the <pre>dearmor</pre> operation using PGPainless.
 */
public class DearmorImpl implements Dearmor {

    @Nonnull
    @Override
    public Ready data(@Nonnull InputStream data) {
        InputStream decoder;
        try {
            decoder = PGPUtil.getDecoderStream(data);
        } catch (IOException e) {
            throw new SOPGPException.BadData(e);
        }
        return new Ready() {

            @Override
            public void writeTo(@Nonnull OutputStream outputStream) throws IOException {
                BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(outputStream);
                Streams.pipeAll(decoder, bufferedOutputStream);
                bufferedOutputStream.flush();
                decoder.close();
            }
        };
    }
}

File: pgpainless-sop/src/main/kotlin/org/pgpainless/sop/DearmorImpl.kt
// SPDX-FileCopyrightText: 2024 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.sop

import java.io.BufferedOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import org.bouncycastle.openpgp.PGPUtil
import org.bouncycastle.util.io.Streams
import sop.Ready
import sop.exception.SOPGPException
import sop.operation.Dearmor

/** Implementation of the `dearmor` operation using PGPainless. */
class DearmorImpl : Dearmor {

    override fun data(data: InputStream): Ready {
        val decoder =
            try {
                PGPUtil.getDecoderStream(data)
            } catch (e: IOException) {
                throw SOPGPException.BadData(e)
            }

        return object : Ready() {
            override fun writeTo(outputStream: OutputStream) {
                BufferedOutputStream(outputStream).use {
                    Streams.pipeAll(decoder, it)
                    it.flush()
                    decoder.close()
                }
            }
        }
    }
}
